compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -0,0 +1,405 @@
+use crate::MirPass;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_index::IndexVec;
+use rustc_middle::middle::resolve_bound_vars::Set1;
+use rustc_middle::mir::visit::{PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::adjustment::PointerCast;
+use rustc_middle::ty::{ParamEnv, TyCtxt};
+use rustc_session::Session;
+use smallvec::SmallVec;
+
+pub struct InstCombine;
+
+impl<'tcx> MirPass<'tcx> for InstCombine {
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.mir_opt_level() >= 1
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        let param_env = tcx.param_env_reveal_all_normalized(body.source.def_id());
+        while did_optimization(body, tcx, param_env) {}
+    }
+}
+
+struct Context<'tcx, 'a> {
+    tcx: TyCtxt<'tcx>,
+    local_decls: &'a LocalDecls<'tcx>,
+    param_env: ParamEnv<'tcx>,
+}
+
+// Combine:
+// _2 = &_1;
+// _3 = *_2;
+// Into:
+// _3 = _1;
+fn combine_ref_deref<'tcx, 'a>(
+    cx: &Context<'tcx, 'a>,
+    _temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Ref(_, _, first_place) = temp_rvalue else { return None; };
+    let Rvalue::Use(final_operand) = final_rvalue else { return None; };
+
+    let Some(second_place) = final_operand.place() else { return None; };
+    if second_place.projection.get(0) != Some(&ProjectionElem::Deref) {
+        return None;
+    }
+
+    let new_place = first_place.project_deeper(&second_place.projection[1..], cx.tcx);
+
+    if new_place == *final_place {
+        Some(StatementKind::Nop)
+    } else {
+        Some(StatementKind::Assign(Box::new((*final_place, Rvalue::Use(Operand::Copy(new_place))))))
+    }
+}
+
+// Combine:
+// _2 = &mut _1;
+// _3 = &raw mut *_2;
+// Into:
+// _3 = &raw mut _1;
+fn combine_ref_addressof<'tcx, 'a>(
+    cx: &Context<'tcx, 'a>,
+    _temp_place: &Place<'tcx>,
+    temp_rvalue: &Rvalue<'tcx>,
+    final_place: &Place<'tcx>,
+    final_rvalue: &Rvalue<'tcx>,
+) -> Option<StatementKind<'tcx>> {
+    let Rvalue::Ref(_, _, first_place) = temp_rvalue else { return None; };
+    let Rvalue::AddressOf(mutability, second_place) = final_rvalue else { return None; };
+
+    if second_place.projection.get(0) != Some(&ProjectionElem::Deref) {
+        return None;
+    }
+
+    let new_place = first_place.project_deeper(&second_place.projection[1..], cx.tcx);
+
+    Some(StatementKind::Assign(Box::new((*final_place, Rvalue::AddressOf(*mutability, new_place)))))
+}
+
+// Combine:
+// _2 = _1.a;
+// _3 = _2.b;
+// Into:
+// _3 = _1.a.b;
```

Comment:
Likewise, could you add a correctness explanation. IMO, this boils down to the fact that we only consider `_2` as a value and not as a place.